### DISCUSSION TITLE
Need a Little Grammar Help

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#1041] [Q&A] [answered]
I'm working on a grammar that needs to support specification of HTTP paths in a similar manner to OpenAPI.  I almost have it working, but I can't get the parser to recognize the `SEGMENT` rule as shown in [this playground session](https://langium.org/playground?grammar=OYJwhgthYgBACmALgCwM4CgMFMB2SQBPBZFAWQHsATbAGwC4NZnYAHUtAagF5FUBBJAQCWAIwCuSbACoA3Fj4pBIiVMYtYAcgDa7VJq30DelNwCqAJQAyWgDQHNEbKmqbNR2E5dVuZZylcAXU15DABlbABjJGEKXEZNAHoDAAphKQgubgBJDIBKaVgUpIN07EyeXPK8gH5YUKqIRkRwCAB9COAnfFgAHxJUDuwuvCRQlsghkfwEgG8DNGHupF5%2BC34yLQBfEIVSKeXYYVwAM2wQNAGUTuX1WEXplbCAUQBxMmeAOQAVUMsrO7MRbRWK4LKwCIguKcUIaPzeWAgZziEBg%2B4iXDAQFFTTAZwGfqaViSAlaVgUNBIUmaGi0ZzYTR5UJSEAQY5gWgINYbeiwbIAEXqWBQwioNFwsBZbNwHNgAHUwrzEgAdLiJZnnaWygVK7RtMAAWgAXvwDQAtQLaZUAdzagWk6qwUvZnJe7y%2B32xKSsz2%2B32eFj6sH52Ve2W%2BQbC8GeAGFsvwrHlOBgNayXbAfX6A17NGBNAA6fOaI3U-gFotmxkpp2a9MhsOepgsYoABnLmgAnFXU1rXdG4wmcwbqW1qUWg5oAH7UgCkBgAEs8ABqwRdLvLVjDOmWctc5tuFzvUvOHk6l9sAMW7wtF4sltZ3sDIVjaMYA8mQPj8lcqVdIrWgyphPaNTKtIv6OiKYp4PeaaPmEL7vp%2BHo-iqiTaAAesquDKiA9qOkAA&content=NoBwhgLgFgXABAejAIwMYIEIBo4FsCm0A9gCYwDmhAugFA2iSwIDeuAngGpgBOAvpjgLEylCFSA)

I don't know why 'abc' or 'B' are not recognized here as SEGMENT.  

Grammar:

```
grammar Paths

entry PathModel:
    paths+=PathAttribute*;

PathAttribute:
    '[path' ':' path=URL ',' 'method'':' method=Method']';

Section:
'/' (items+=Item)* ('/' items+=Item)? ;

Item:
Param_Segment | Path_Segment;

Param_Segment:
'{' segment=PARAM '}';

Path_Segment infers PathSegment:
  segment=SEGMENT;

URL:
    sections+= Section+;

    Method returns string:
    ('get' | 'put' | 'post' | 'delete');

terminal PARAM: ID ;

hidden terminal WS: /\s+/;

terminal ID: /[_a-zA-Z][\w_]*/;

terminal SEGMENT:
    (LETTER | DIGIT | SPECIAL)+
;

terminal LETTER:
    ('a'..'z' | 'A'..'Z')
;

terminal DIGIT:
    ('0'..'9')
;

terminal SPECIAL:
    ('-' | '_' | '.' | '~' | '%' HEX HEX)
;

terminal HEX:
    ('0'..'9' | 'a'..'f' | 'A'..'F')
;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;

```

**Comment by msujew** [comment, marked as answer]
@ballcoach12 Your terminals are clashing quite heavily. In `/abc/B` both `abc` and `B` get identified as `PARAM` terminals, since that's the first terminal in your terminal list that is able to tokenize these characters. Note that the tokenization and parsing phase are split and the tokenization (lexing) is performed without context (whereas the parsing phase uses context).

In your case, `PARAM`, `ID`, `SEGMENT` and `LETTER` are ambiguous during the tokenization phase. The tokenizer will never read `ID`, since the first-fit algorithm employed there will always read `PARAM` first. A similar thing happens with `SEGMENT` since there's quite an overlap in these token types.

**Reply by ballcoach12** [reply]
Thank you, @msujew! Sometimes I can't see the forest for the trees! 

I revised my grammar to clear up the clutter, and it now works.  Here is the new grammar:

```
grammar Paths

entry PathModel:
    paths+=PathAttribute*;

PathAttribute:
    '[path' ':' path=URL ',' 'method'':' method=Method']';

Section:
'/' (items+=Item)* ('/' items+=Item)? ;

Item:
Param_Segment | Path_Segment;

Param_Segment:
'{' segment=ID '}';

Path_Segment infers PathSegment:
  segment=ID;

URL:
    sections+= Section+;

    Method returns string:
    ('get' | 'put' | 'post' | 'delete');

hidden terminal WS: /\s+/;

terminal ID: /[_a-zA-Z][\w_]*/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```
